titan-compiler/coder.lua

```diff
@@ -197,19 +345,151 @@ generate_program = function(prog, modname)
             end
         end
 
-        table.insert(parts, "}")
-        create_module_table = table.concat(parts, "\n")
+        create_module_table = util.render([[
+            {
+                /* Initialize module table */
+                lua_createtable(L, 0, ${N});
+                ${PARTS}
+            }
+        ]], {
+            N = c_integer(n_exported_functions),
+            PARTS = table.concat(parts, "\n")
+        })
     end
 
     local code = util.render(whole_file_template, {
         MODNAME = modname,
+        N_TOPLEVEL = c_integer(n_toplevel),
+        DEFINE_FUNCTIONS = define_functions,
         INITIALIZE_TOPLEVEL = initialize_toplevel,
         CREATE_MODULE_TABLE = create_module_table,
     })
     return pretty.reindent_c(code)
 end
 
--- Returns (statements, value)
+
+generate_stat = function(stat)
+    local tag = stat._tag
+    if     tag == ast.Stat.Block then
+        local cstatss = {}
+        table.insert(cstatss, "{")
+        for _, inner_stat in ipairs(stat.stats) do
+            local cstats = generate_stat(inner_stat)
+            table.insert(cstatss, cstats)
+        end
+        table.insert(cstatss, "}")
+        return table.concat(cstatss, "\n")
```

Comment:
Do we need to return a string here? We are calling `table.concat` recursively, I'm afraid that this might generate a performance issue.